### ISSUE TITLE
Vulkan: long-prompt (200 tok) prefill logits diverge from CPU on llama (short prompts pass)

### ISSUE THREAD

**Issue by rexlunae** [rexlunae/joshua#81]
## Summary

`tests/backend_matrix.rs` (added for cross-backend regression coverage) found a real Vulkan correctness divergence: the **llama** fixture's **long prefill (200 tokens)** produces materially different logits on the Vulkan path than on CPU, while the **short prefill passes** on the same model.

```
[llama × vulkan] long prefill (200 tok): logit 1 diverges:
  -0.19582216 (vulkan) vs -0.06327148 (cpu)  (tol 0.044014882)
```

Reproduced on the Renoir iGPU (10.0.0.223, RADV, `JOSHUA_VULKAN_NATIVE=1`), `cargo test --release --features vulkan --test backend_matrix`. The OpenCL column of the same matrix (4 arches × the same long prompt) passes on both the B50 and the Renoir, so the divergence is Vulkan-specific.

## What it implicates

Long-sequence attention on Vulkan for the llama arch: the short prompt (5 tokens) matches CPU within rounding, the 200-token prompt does not. Candidate areas: causal-mask shape/handling at longer seq, softmax/reduction over ~200 positions, or the llama KV path. The other arches in the matrix (qwen3moe, deepseek2, deepseek4) did not reach the llama long-prefill check because the matrix asserts llama first (alphabetical fixture order).

## Repro

```
# on an OpenCL/Vulkan host with the fixture models
JOSHUA_VULKAN_NATIVE=1 cargo test --release --features vulkan --test backend_matrix -- --nocapture
```

The matrix prints `matrix: <arch> × <backend> OK` per passing cell; this cell fails. Short-prompt parity for the same fixture passes (vulkan_llama_matches_cpu is green), so the bug is length-dependent.

## Suggested next steps

- [ ] Bisect by prompt length (e.g. 32/64/128/200) to find the threshold.
- [ ] Check the Vulkan causal mask against the CPU mask at 200 positions.
- [ ] Compare softmax_last/reduce outputs over 200-length rows device vs CPU.
- [ ] If a single native kernel is wrong, `JOSHUA_OPENCL_NATIVE_DENY`-style deny for Vulkan (if added) can name it; otherwise fix the kernel.

**Comment by rexlunae** [comment]
**Isolated to `k_softmax_last` — with element-level evidence.** New info from `examples/vk_op_probe.rs` (graded-length op probe, CPU vs Vulkan on RADV/Renoir):

- `softmax_last` is **correct for row length ≤ 64 and wrong for ≥ 72** on this device (8/32/48/64 PASS; 72/80/88/96/104/128/160/200 FAIL, worst errors ~2400).
- `rms_norm`, `rope` (broadcast sin/cos), `broadcast mul`, and the **input upload** (device input read back and compared element-exact) are all clean at every length — the divergence is entirely inside the softmax kernel.
- Element diagnosis of a failing row (cols=96): elements 0..63 are exactly **0.492×** CPU and elements 64..95 are exactly **15.52×** CPU — two clean scale factors split at the work-group boundary, stable across runs. The row sums to ~56 instead of 1.
- A first fix attempt (write each element exactly once fully normalized — no raw-exp-then-scale over global memory) **did not change the failure**, which rules out the global-memory write/scale visibility hazard and points at the kernel's two-phase shared-memory reduction interacting with `cols > WG` stride iterations (or a WG/launch mismatch).

**Remaining diagnostic step:** print the device's actual `Limits::wg` at the softmax launch and the per-element values of `sh` mid-kernel; then either fix the stride/reduction interplay or cap `wg` for this kernel so every lane covers the whole row in one iteration.

**Scope note:** this also explains #81's original symptom — llama's attention softmax over [seq, seq] rows of length ≥ 72 on Vulkan produces the divergent logits.

**Comment by rexlunae** [comment]
**SYCL backend (Intel oneAPI/DPC++) phase 1 landed** (`1056fc7`..`f0fbae7` on main): the other agent's C++ bridge (68 kernels ported 1:1 from OpenCL, dispatch.inc typed launchers, bridge.cpp C ABI) is now committed alongside the Rust loading module (`sycl_backend/mod.rs` — dlopen loader, function-pointer resolution via libloading, `SyclDevice` with alloc/read/write/launch, `ArgBuilder` with correct per-arg sizes) and `tests/sycl_bridge_tests.rs` (kernel parity tests, the same role the OpenCL parity tests play).

**Validated on the Arc Pro B50 (10.0.0.13, DPC++ nightly 7.2.0):**
- ✅ `sycl_affine_matches_cpu` — PASS
- ✅ `sycl_binary_add_matches_cpu` — PASS
- ✅ bridge loads, context opens, device info reads, alloc/write/read/launch all work
- 🔴 gemm (16×16 tile) — SIGSEGV, likely the 3D nd_range dimension reversal (OpenCL x/y/z → SYCL z/y/x)
- 🔴 softmax, rmsnorm, gemv-nt, qgemv, hembed — `UR_RESULT_ERROR_OUT_OF_RESOURCES` (work-group / local-memory sizing)

Both failure classes are first-run-on-hardware kernel bugs, exactly what the parity tests are designed to catch. The fixes are straightforward SYCL nd_range debugging once someone picks this up.

**Comment by rexlunae** [comment]
**SYCL backend status update:**

Phase 1 infrastructure landed and validated on the Arc Pro B50 (`1056fc7`..`0b09a5e` on main):

**Working:**
- Bridge loads, context opens, `Intel(R) Arc(TM) Pro B50 Graphics (15488 MiB)` detected via DPC++ nightly 7.2.0
- `sycl_affine_matches_cpu` — ✅ PASS (correct results on hardware)
- `sycl_binary_add_matches_cpu` — ✅ PASS
- Device alloc/write/read/launch/finish all functional through the Rust bridge

**Not yet working (first-run kernel bugs):**
- `sycl_unary_exp_matches_cpu` — SIGSEGV (the simplest 1D kernel crashes)
- `sycl_softmax_matches_cpu` — SIGSEGV
- `sycl_rmsnorm_matches_cpu` — SIGSEGV
- `sycl_qgemv_q8_0_matches_cpu` — SIGSEGV
- `sycl_hembed_matches_cpu` — SIGSEGV (wrong kernel argument count)
- `sycl_gemm_matches_cpu` — SIGSEGV (3D nd_range reversal, explicitly gated)

All the non-elementwise kernels crash on first execution. The `k_unary_c` kernel is particularly telling: it's a trivial 1D elementwise op (`o[i] = f(x[i])`) with no shared memory, no tile logic, and the launch dimensions are verified correct. Its SIGSEGV suggests the DPC++ compiler generates GPU code that writes out of bounds — possibly because the `local_accessor<uint64_t, 1>` scratch pointer (which `k_unary_c` never dereferences) is invalid, or because the `Kernels{item, scratch}` constructor's aggregate initialization has a compiler bug.

The affine and binary kernels, which have the same structure but one more argument, work correctly. This suggests the issue is specifically in how the DPC++ compiler handles the `Kernels{item, scratch}` aggregate initialization for certain kernel signatures.

**Debugging approach:** the same element-level diagnosis used for the Vulkan softmax (#81's first fix) — isolate which kernel parameter causes the crash by bisecting the kernel's argument count (start with a 1-arg trivial kernel, add args until it crashes).

The bridge infrastructure, feature gates, test framework, and kernel port are all committed and pushed. Phase 2 (a full BackendStorage/BackendDevice integration wired into Device) is blocked on these kernel bugs.

**Comment by rexlunae** [comment]
**SYCL backend finished — all 8 kernel parity tests pass on the Arc Pro B50.**

Final state (`c214c19` on main): the bridge, Rust launchers, and `tests/sycl_bridge_tests.rs` all work on real DPC++ nightly 7.2.0 hardware:

```
SYCL device: Intel(R) Arc(TM) Pro B50 Graphics (15488 MiB)
  affine OK / unary_exp OK / binary_add OK / softmax OK
  rmsnorm OK / gemm OK / qgemv_q8_0 OK / hembed OK
test result: ok. 1 passed; 0 failed
```

The three real bugs found and fixed along the way:

1. **`ArgBuilder` dangling pointers (the segfault root cause)** — the launch-argument builder took `arena.as_ptr()` during every push; later pushes reallocated the Vec and dangled every earlier `SyclArg.data`. The stale heap usually still held the old bytes (so each kernel passed when it ran first and crashed whatever ran second — the crash "followed the second test" regardless of kernel). Fixed by recording (offset, size) pairs and materialising the descriptors once after the arena is final.

2. **`k_gemm`'s Bs tile overlapped As in shared memory** — the OpenCL original declares two separate `__local` arrays; the SYCL port replaced them with one scratch pointer plus a manual Bs offset of `TK * TM + 1` = 1025 floats while As actually occupies `TK * (TM + 1)` = 1040. Bs started 15 floats early, clobbering As's last row — gemm rows ≥ 50 wrong. Proven at the element level (identity-encoded dumps: thread ids, M/N/K, and the writer's gm/ty were all correct inside the kernel; `As[15][50] == Bs[0][0] == B[0] == -2.0`). Fixed the offset to `TK * (TM + 1)`.

3. **`k_hembed` argument count and semantics** — the kernel takes the backend's fault checker (11 args, matching the OpenCL launcher), and the launcher must take `n_ids` and compute `n = n_ids * k` internally (the caller was passing `n_ids` as the element total, so only one row was gathered and the rest of the readback was stale memory).

Infrastructure fixes along the way: WG 256→64 (Intel register pressure / OUT_OF_RESOURCES), serial test wrapper + shared device + global op mutex (DPC++ runtime is not thread-safe for concurrent launches), libloading function-pointer resolution (no link-time undefined symbols), `sycl::native` math in the unary op table (precise DPC++ expansions blew the register budget — required for the kernel to launch at all on this device).
